File: src/avr/isp.rs
use std::rc::Rc;

use rusb::Result;

use crate::ll::ATPR;

use super::Program;

pub struct ISP {
    handle: Rc<ATPR>,
}

impl From<Rc<ATPR>> for ISP {
    fn from(handle: Rc<ATPR>) -> Self {
        Self { handle }
    }
}

impl Program for ISP {
    fn connect(&self) -> Result<()> {
        log::debug!("Connecting to device");
        self.handle.write(0x00, 0x00, 0x00, &[]).map(|_| ())
    }

    fn begin(&self, isp_speed: u8) -> Result<bool> {
        assert!(isp_speed < 14);
        log::debug!("Begin ISP transfer with speed grade {}", isp_speed);

        let isp_speed = match isp_speed {
            0 => 0b1000, // f_osc / 4
            1 => 0b1101, // f_osc / 8
            2 => 0b1001, // f_osc / 16
            3 => 0b1110, // f_osc / 32
            4 => 0b1010, // f_osc / 64
            5 => 0b1011, // f_osc / 128
            _ => isp_speed - 6,
        };

        let mut buf = [0u8];
        self.handle
            .read(0x00, isp_speed as _, 0x00, &mut buf)
            .map(|_| buf[0] == 0x01)
            .map_err(|e| {
                log::error!("Can't begin: target doesn't answer.");
                e
            })
    }

    fn end(self) -> Result<()> {
        todo!()
    }

    fn chip_erase(&self) -> Result<bool> {
        todo!()
    }

    fn is_ready(&self) -> Result<bool> {
        todo!()
    }

    fn latch_data(&self) -> Result<bool> {
        todo!()
    }

    fn read_flash(&self, address: usize) -> Result<u8> {
        todo!()
    }

    fn write_flash(&self, address: usize, data: u8) -> Result<bool> {
        todo!()
    }

    fn load_flash_page(&self, address: usize, data: &[u8]) {
        todo!()
    }

    fn flush_flash_page(&self) -> Result<bool> {
        todo!()
    }

    fn read_eeprom(&self, address: usize) -> Result<u8> {
        todo!()
    }

    fn write_eeprom(&self, address: usize, data: u8) -> Result<bool> {
        todo!()
    }
}

File: src/avr/isp.cpp
/* ATPRsup - Software for ATPR
 * Copyright (C) 2021-2022 45gfg9
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "avr/isp.h"
#include <cassert>

using namespace AVR;

// TODO

ISP::ISP(ATPR *handle, uint8_t ispSpeed) : Program(handle) {
    assert(ispSpeed < 14);

    const uint8_t map[6] = {0b1000, 0b1101, 0b1001, 0b1110, 0b1010, 0b1011};
    this->ispSpeed = ispSpeed < 6 ? map[ispSpeed] : ispSpeed - 6;

    handle->write(0x00, 0x0000, 0x0000, nullptr, 0);
}

ISP::~ISP() = default;

bool ISP::begin() {
    uint8_t resp;

    handle->read(0x00, ispSpeed, 0x0000, &resp, 1);

    return resp;
}

bool ISP::chipErase() {
    return false;
}

bool ISP::isReady() {
    return false;
}

bool ISP::latchData() {
    return false;
}

uint8_t ISP::readFlash(uint32_t address) {
    return 0;
}

bool ISP::writeFlash(uint32_t address, uint8_t data) {
    return false;
}

bool ISP::loadFlashPage(uint32_t address, const uint8_t *data, uint16_t pageSize) {
    return false;
}

bool ISP::flushFlashPage() {
    return false;
}

uint8_t ISP::readEeprom(uint32_t address) {
    return 0;
}

bool ISP::writeEeprom(uint32_t address, uint8_t data) {
    return false;
}

uint8_t ISP::readFuse(uint8_t address) {
    return 0;
}

bool ISP::writeFuse(uint8_t address, uint8_t data) {
    return false;
}

uint8_t ISP::readLockBits() {
    return 0;
}

const uint8_t *ISP::readSignature() {
    uint8_t buf[4];
    for (uint8_t i = 0; i < 3; i++) {
        handle->read(0x01, 0x0030, i, buf, 4);

        sigBuf[i] = buf[3];
    }

    return sigBuf;
}
